components/modules/matrix.c

```diff
@@ -0,0 +1,528 @@
+                          .mode = GPIO_MODE_INPUT,
+                          .pull_up_en = GPIO_PULLUP_ENABLE,
+                          .pull_down_en = GPIO_PULLDOWN_DISABLE,
+                          .intr_type = intr};
+
+  return gpio_config(&config);
+}
+
+static esp_err_t set_gpio_mode_output(int pin) {
+  gpio_config_t config = {.pin_bit_mask = 1LL << pin,
+                          .mode = GPIO_MODE_OUTPUT_OD,
+                          .pull_up_en = GPIO_PULLUP_DISABLE,
+                          .pull_down_en = GPIO_PULLDOWN_DISABLE
+                        };
+
+  return gpio_config(&config);
+}
+
+static void set_columns(DATA *d, int level) {
+  for (int i = 0; i < d->column_count; i++) {
+    gpio_set_level(d->columns[i], level);
+  }
+}
+
+static void initialize_pins(lua_State *L, DATA *d) {
+  for (int i = 0; i < d->column_count; i++) {
+    if (set_gpio_mode_output(d->columns[i]) != ESP_OK) {
+      luaL_error(L, "Unable to configure pins");
+    }
+  }
+
+  set_columns(d, 0);
+
+  for (int i = 0; i < d->row_count; i++) {
+    if (set_gpio_mode_input(d->rows[i], GPIO_INTR_NEGEDGE) != ESP_OK) {
+      luaL_error(L, "Unable to configure pins");
+    }
+  }
+}
+
+static void disable_row_interrupts(DATA *d) {
+  for (int i = 0; i < d->row_count; i++) {
+    gpio_set_intr_type(d->rows[i], GPIO_INTR_DISABLE);
+  }
+}
+
+// Just takes the channel number. Cleans up the resources used.
+static int matrix_close(DATA *d) {
+  if (!d) {
+    return 0;
+  }
+
+  disable_row_interrupts(d);
+
+  for (int i = 0; i < d->row_count; i++) {
+    gpio_isr_handler_remove(d->rows[i]);
+  }
+
+  for (int i = 0; i < d->column_count; i++) {
+    set_gpio_mode_input(d->columns[i], GPIO_INTR_DISABLE);
+  }
+
+  return 0;
+}
+
+// Character returned is 0 .. max if pressed. -1 if not.
+static int matrix_get_character(DATA *d)
+{
+  set_columns(d, 1);
+  disable_row_interrupts(d);
+
+  int character = -1;
+
+  // We are either waiting for a negative edge (keypress) or a positive edge
+  // (keyrelease)
+
+  for (int i = 0; i < d->column_count && character < 0; i++) {
+    gpio_set_level(d->columns[i], 0);
+
+    for (int j = 0; j < d->row_count && character < 0; j++) {
+      if (gpio_get_level(d->rows[j]) == 0) {
+        // We found a keypress
+        character = j * d->column_count + i;
+      }
+    }
+
+    gpio_set_level(d->columns[i], 1);
+  }
+
+  return character;
+}
+
+static void matrix_queue_character(DATA *d, int character)
+{
+  // If character is >= 0 then we have found the character -- so send it.
+
+  if ((d->state == WAITING_FOR_PRESS && character >= 0) || (d->state == WAITING_FOR_RELEASE && character < 0)) {
+    if (character >= 0) {
+      character++;
+      d->last_character = character;
+    } else {
+      character = -d->last_character;
+    }
+
+    if (HAS_QUEUE_SPACE(d)) {
+      QUEUE_IT(d, character);
+      if (!d->task_queued) {
+        if (task_post_medium(tasknumber, (task_param_t)d)) {
+          d->task_queued = 1;
```

Comment:
I believe `d->task_queued` needs to be mutex protected, as it gets written by both the GPIO ISR/task as well as the NodeMCU task.